zebra-network/src/peer/handshake.rs

```diff
@@ -464,19 +464,39 @@ where
                                 let (tx, rx) = oneshot::channel();
                                 let request = Request::Ping(Nonce::default());
                                 tracing::trace!(?request, "queueing heartbeat request");
-                                if server_tx
-                                    .send(ClientRequest {
-                                        request,
-                                        tx: tx.into(),
-                                        span: tracing::Span::current(),
-                                    })
-                                    .await
-                                    .is_err()
-                                {
-                                    tracing::trace!(
-                                        "error sending heartbeat request, shutting down"
-                                    );
-                                    return;
+                                match server_tx.try_send(ClientRequest {
+                                    request,
+                                    tx: tx.into(),
+                                    span: tracing::Span::current(),
+                                }) {
+                                    Ok(()) => {
+                                        match server_tx.flush().await {
+                                            Ok(()) => {}
+                                            Err(e) => {
+                                                // TODO: we can't get the client request for this failure,
+                                                // so we can't ensure the invariant holds
+                                                panic!("flushing client request failed: {:?}", e);
```

Comment:
We panic if flush fails, but the previous code just closed the connection.

Here are our options for fixing this issue:
* does flush ever fail?
  * if it never fails, we don't actually need to fix anything
* can we cleanup the ClientRequest some other way?
  * is there a `flush` variant that returns the message on error?
  * is there a `try_send_and_flush` function?

Reply:
@yaahc the problem seems to be that most futures APIs aren't designed to return messages on failure.

We could auto-send ConnectionClosed on drop in this one very specific case, and leave all the rest of the code to panic on drop?

Reply:
I think applying `MustUseOneshotSender` after the `ClientRequest` is successfully received resolves this issue, because a failed `flush` means that the request was never received.

Reply:
> I think applying `MustUseOneshotSender` after the `ClientRequest` is successfully received resolves this issue, because a failed `flush` means that the request was never received.

I agree, this is precisely what the `InProgressClientRequest` would do by encoding "applying the MustUseOneshotSender` to a `ClientRequest` once it is received" in the type system.